subs/pantry/src/Pantry/SQLite.hs

```diff
@@ -1,47 +1,93 @@
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE OverloadedStrings #-}
 module Pantry.SQLite
-  ( P.Storage
+  ( Storage (..)
   , initStorage
-  , withStorage
   ) where
 
 import RIO hiding (FilePath)
-import qualified Pantry.Types as P
 import Database.Persist.Sqlite
 import RIO.Orphans ()
 import Path (Path, Abs, File, toFilePath, parent)
 import Path.IO (ensureDir)
-import Data.Pool (destroyAllResources)
-import Pantry.Types (PantryException (MigrationFailure))
+import Pantry.Types (PantryException (MigrationFailure), Storage (..))
+import System.FileLock (withFileLock, withTryFileLock, SharedExclusive (..))
 
 initStorage
   :: HasLogFunc env
   => Text
   -> Migration
   -> Path Abs File -- ^ storage file
-  -> (P.Storage -> RIO env a)
+  -> (Storage -> RIO env a)
   -> RIO env a
 initStorage description migration fp inner = do
   ensureDir $ parent fp
-  bracket
-    (createSqlitePoolFromInfo (sqinfo False) 1)
-    (liftIO . destroyAllResources) $ \pool -> do
-    migrates <- wrapMigrationFailure $ runSqlPool (runMigrationSilent migration) pool
-    forM_ migrates $ \mig -> logDebug $ "Migration executed: " <> display mig
-  bracket
-    (createSqlitePoolFromInfo (sqinfo True) 1)
-    (liftIO . destroyAllResources) $ \pool -> inner (P.Storage pool)
+
+  migrates <- withWriteLock fp $ wrapMigrationFailure $
+    withSqliteConnInfo (sqinfo True) $ runReaderT $
+    runMigrationSilent migration
+  forM_ migrates $ \mig -> logDebug $ "Migration executed: " <> display mig
+
+  inner $ Storage
+    { withStorage_ = withWriteLock fp . withSqliteConnInfo (sqinfo False) . runSqlConn
+    , withWriteLock_ = id
+    }
   where
     wrapMigrationFailure = handle (throwIO . MigrationFailure description fp)
```

Comment:
This only catches `SqliteException`, but I noticed that in some cases persistent calls `error`.  For example [if there's an unsafe migration](https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist/Sql/Migration.hs#L103) that doesn't get caught here and so the error doesn't give any useful advice for an end-user.